source/common/network/connection_impl.cc

```diff
@@ -886,11 +904,18 @@ ClientConnectionImpl::ClientConnectionImpl(
 }
 
 void ClientConnectionImpl::connect() {
-  ENVOY_CONN_LOG(debug, "connecting to {}", *this,
-                 socket_->connectionInfoProvider().remoteAddress()->asString());
+
+  ENVOY_CONN_LOG(debug, "connecting to {}, see bind_error = {}", *this,
+                 socket_->connectionInfoProvider().remoteAddress()->asString(), bind_error_);
+  // TODO(lambdai): we should avoid calling connect if the bind() fails already.
   const Api::SysCallIntResult result =
       socket_->connect(socket_->connectionInfoProvider().remoteAddress());
   if (result.return_value_ == 0) {
+    if (immediate_error_event_.has_value()) {
+      ASSERT(*immediate_error_event_ == ConnectionEvent::LocalClose);
+    } else {
+      immediate_error_event_ = ConnectionEvent::Connected;
+    }
```

Comment:
Alternate thing you could consider here:

if (immediate_error_event_ != ConnectionEvent::Connected) {
  // force a wakeup
  activate(Write);
}

This change will allow you to remove the activate(Write) in the constructor while still delivering the immediate_error_event_ in a timeline fashion.